### ISSUE TITLE
The dependency tree should be fully resolved

### ISSUE THREAD

**Issue by durka** [Nemo157/cargo-lichking#10]
I don't know how to fix this error:

```
$ cargo build --no-default-features
<snip>
$ cargo lichking list
warning: IANAL: This is not legal advice and is not guaranteed to be correct.
thread 'main' panicked at 'The dependency tree should be fully resolved', src/libcore/option.rs:785
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

How can the dependencies not be resolved? I just successfully built the program.

The code is [here](https://github.com/haptics-nri/nri/tree/error-handling) if you need it but will take a long time to compile.

OSX 10.9.5, rustc 1.18.0-nightly (63c77214c 2017-04-24), cargo 0.19.0-nightly (8326a3683 2017-04-19), cargo-lichking 0.4.1

**Comment by Nemo157** [comment]
Interesting, the error that is being hit is here: https://github.com/Nemo157/cargo-lichking/blob/master/src/load.rs#L31.

I'm guessing this is something to do with not handling optional packages correctly. I'll take a look and see if I can improve the error message and fix this.

**Comment by Nemo157** [comment]
Threw in a little bit better error message and got:

```
→ ../cargo-lichking/target/debug/cargo-lichking lichking list
warning: IANAL: This is not legal advice and is not guaranteed to be correct.
thread 'main' panicked at 'The dependency tree should be fully resolved, failed to find hprof v0.1.3 (file:///Users/Nemo157/sources/nri-error-handling/overrides/hprof)', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libcore/option.rs:715
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

So it looks like there's something wrong with the override handling.

**Comment by Nemo157** [comment]
Ok, found one potential solution, ignore replacements

```diff
diff --git src/load.rs src/load.rs
index 54a58d3..0121fb7 100644
--- src/load.rs
+++ src/load.rs
@@ -19,7 +19,7 @@ pub fn resolve_packages(
     while let Some(id) = to_check.pop() {
         if let Ok(package) = packages.get(id) {
             if result.insert(package) {
-                let deps = resolve.deps(id);
+                let deps = resolve.deps_not_replaced(id);
                 for dep_id in deps {
                     let dep = package.dependencies().iter()
                         .find(|d| d.matches_id(dep_id))
```

I'm not certain this is what we actually want to do though, what if someone changes the license in the replacement?

**Comment by durka** [comment]
It looks like I have changed the hprof version from 0.1.2 to 0.1.3 in my submodule. Not sure if that would change the error or not.

**Comment by Nemo157** [comment]
Nope, I had a previous error message until I also edited that:

```
error: no matching package for override `https://github.com/rust-lang/crates.io-index#hprof:0.1.3` found
location searched: file:///Users/Nemo157/sources/nri-error-handling/overrides/hprof
version required: = 0.1.3
```

I think it should be fine to just ignore replacements in this case. There's probably no chance of someone changing the license in a replacement legally anyway. The only real potential issue I can see is that a replacement pulls in an extra dependency, but I can add a warning about that if it sees a replacement.

**Comment by durka** [comment]
But then you'd want a warning if someone *did* change the license in a replacement... Cargo must know where the replaced packages are.

**Comment by Nemo157** [comment]
Ah, no, I just realised how this code is actually working. It's only using the package id it gets here to look up the dependency in the dependers manifest so it can ignore build and development dependencies. That's why using the replaced id fails, the dependency is specified via the original id. So this change is correct for the most part, but then when it goes to load the dependents package details it needs to load the replaced package.

**Comment by Nemo157** [comment]
Fix will be released in 0.4.2 once Travis gets round to it, thanks for reporting 😃
